libcaf_core/caf/message_builder.test.cpp

```diff
@@ -0,0 +1,191 @@
+    }
+    return inspect(f, *x);
+  }
+}
+
+CAF_BEGIN_TYPE_ID_BLOCK(counted_int_block, caf::first_custom_type_id)
+
+  CAF_ADD_TYPE_ID(counted_int_block, (counted_int))
+  CAF_ADD_TYPE_ID(counted_int_block, (counted_int_ptr))
+
+CAF_END_TYPE_ID_BLOCK(counted_int_block)
+
+TEST("message_builder can build messages incrementally") {
+  message_builder builder;
+  SECTION("a default-constructed builder is empty") {
+    check(builder.empty());
+    check(builder.to_message().empty());
+    check_eq(builder.size(), 0u);
+  }
+  SECTION("append() adds values to a builder") {
+    print_debug("after adding 1, the message is (1)");
+    {
+      builder.append(int32_t{1});
+      auto msg = builder.to_message();
+      check_eq(builder.size(), 1u);
+      check_eq(msg.types(), make_type_id_list<int32_t>());
+      check_eq(msg.get_as<int>(0), 1);
+    }
+    print_debug("after adding [2, 3], the message is (1, 2, 3)");
+    {
+      std::vector<int32_t> xs{2, 3};
+      builder.append(xs.begin(), xs.end());
+      check_eq(builder.size(), 3u);
+      auto msg = builder.to_message();
+      check_eq(msg.types(), make_type_id_list<int32_t, int32_t, int32_t>());
+      check_eq(msg.get_as<int>(0), 1);
+      check_eq(msg.get_as<int>(1), 2);
+      check_eq(msg.get_as<int>(2), 3);
+    }
+  }
+  SECTION("append_from() adds values from another message") {
+    std::vector<int32_t> xs{1, 2, 3};
+    builder.append(xs.begin(), xs.end());
+    auto msg = builder.to_message();
+    message_builder builder2;
+    check(builder2.empty());
+    check(builder2.to_message().empty());
+    SECTION("builder appends the values when first < msg.size()") {
+      builder2.append_from(msg, 0, msg.size());
+      auto msg2 = builder2.to_message();
+      check_eq(msg2.types(), make_type_id_list<int32_t, int32_t, int32_t>());
+      check_eq(msg2.get_as<int>(0), 1);
+      check_eq(msg2.get_as<int>(1), 2);
+      check_eq(msg2.get_as<int>(2), 3);
+    }
+    SECTION("builder does not append the value when first >= msg.size()") {
+      builder2.append_from(msg, 4, msg.size());
+      auto msg2 = builder2.to_message();
+      check_eq(msg2.size(), 0u);
+    }
+    SECTION("builder appends the values from first index to msg.size()") {
+      builder2.append_from(msg, 2, msg.size());
+      auto msg2 = builder2.to_message();
+      check_eq(msg2.size(), 1u);
+      check_eq(msg2.get_as<int>(0), 3);
+    }
+  }
+  SECTION("move_to_message moves all elements into a message") {
+    builder.append("foo");
+    builder.append(make_counted<counted_int>(42));
+    auto msg = builder.move_to_message();
+    if (check_eq(msg.types(),
+                 make_type_id_list<std::string, counted_int_ptr>())) {
+      check_eq(msg.get_as<std::string>(0), "foo");
+      const auto& ci = msg.get_as<counted_int_ptr>(1);
+      check_eq(ci->value(), 42);
+      check_eq(ci->get_reference_count(), 1u);
+    }
+  }
+  SECTION("move_to_message copies message appended with append_form") {
```

Comment:
This test overall became a bit confusing by adding too many sections to it. At the very least, I'd like to see the `move_to_message` tests move out of this test. Since these tests are a bit more involved, I think they would benefit from structuring them as scenarios:

```
SCENARIO: move_to_message move-constructs message elements
  GIVEN: a builder object
    WHEN: appending primitive values
      THEN: move_to_message copies values
        - builder.append with integers
    WHEN: appending movable values
      THEN: move_to_message moves values into place
        - builder.append with counted_int_ptr
  GIVEN: a builder object and a message
    WHEN: appending from a the message with multiple references
      THEN: move_to_message copies the value from the message
        - msg = make_message(make_counted<counted_int> ...
        - use append_from and hold onto the reference in `msg`
    WHEN: appending from a message with no other references
      THEN: move_to_message moves the value from the message
        - same as above but use append_from(make_message(... to have no extra ref to the message
```

Should make it more obvious what's actually being tested here, don't you think?

Reply:
Yes, I have updated the restructured tests and added an additional test according to the given scenarios.